File: demo-test/src/test/java/io/fries/demo/test/cucumber/world/scenario/ScenarioId.java
package io.fries.demo.test.cucumber.world.scenario;

import io.cucumber.java.Scenario;

import java.text.Normalizer;

import static java.util.Objects.requireNonNull;

public record ScenarioId(String value) {

    private static final String FEATURES_BASE_DIRECTORY = "features/";

    public ScenarioId {
        requireNonNull(value, "value cannot be null");
    }

    public static ScenarioId from(final Scenario scenario) {
        final var scenarioUri = scenario.getUri().toString();
        final var scenarioDirectories = scenarioUri.substring(
                scenarioUri.indexOf(FEATURES_BASE_DIRECTORY) + FEATURES_BASE_DIRECTORY.length(),
                scenarioUri.lastIndexOf('.')
        );
        final var scenarioName = Normalizer
                .normalize(scenario.getName(), Normalizer.Form.NFD)
                .toLowerCase()
                .replaceAll(" ", "-")
                .replaceAll("[^a-z\\d\\-]", "");
        final var scenarioPath = String.format("%s/%s", scenarioDirectories, scenarioName);

        return new ScenarioId(scenarioPath);
    }

    @Override
    public String toString() {
        return value;
    }
}

File: demo-test/src/test/kotlin/io/fries/demo/test/cucumber/world/scenario/ScenarioId.kt
package io.fries.demo.test.cucumber.world.scenario

import io.cucumber.java.Scenario
import java.text.Normalizer
import java.util.*

class ScenarioId(private val value: String) {

    companion object {

        private const val FEATURES_BASE_DIRECTORY = "features/"

        fun from(scenario: Scenario): ScenarioId {
            val scenarioUri = scenario.uri.toString()
            val scenarioDirectories = scenarioUri.substring(
                scenarioUri.indexOf(FEATURES_BASE_DIRECTORY) + FEATURES_BASE_DIRECTORY.length,
                scenarioUri.lastIndexOf('.')
            )
            val scenarioName = Normalizer
                .normalize(scenario.name, Normalizer.Form.NFD)
                .lowercase(Locale.getDefault())
                .replace(" ".toRegex(), "-")
                .replace("[^a-z\\d\\-]".toRegex(), "")

            return ScenarioId("$scenarioDirectories/$scenarioName")
        }
    }

    override fun toString(): String = value
}